### PR TITLE
Update memory manager header file to avoid compiling error

### PR DESCRIPTION
**Pull request by qingsenzhang**

### CODE REVIEW

**Review by Sotskin — CHANGES_REQUESTED** [review]
Please add this API:
Get Memory Info: returns total memory and the biggest size of free memory that can be allocated at the time. (See  "cudaMemGetInfo")

File: include/mxnet/mem_mgr.h [hunk 1]
```diff
@@ -1,17 +1,79 @@
...
+  std::mutex mutex_;
   public:
     static MemoryManager* Get();
     static std::shared_ptr<MemoryManager> _GetSharedRef();
     ~MemoryManager();
-    cudaError_t Malloc(void** devptr, size_t size);
-    cudaError_t Free(void* devptr);
+    cudaError_t Malloc(void*& devptr, size_t size, int deviceIdx);
+    cudaError_t Free(void* devptr, int deviceIdx);
+    cudaError_t Memcpy(void* dst, const void* src, size_t count, enum cudaMemcpyKind kind);
```
Comment (Sotskin) [inline include/mxnet/mem_mgr.h]: Also, this function should take in the device id of gpu to copy to/from

**Review by Sotskin — CHANGES_REQUESTED** [review]

File: include/mxnet/mem_mgr.h [hunk 2]
```diff
@@ -1,17 +1,81 @@
 #include <cuda_runtime_api.h>
 #include <iostream>
 #include <memory>
+#include <mutex>
 #include <stdio.h>
 
 namespace mxnet {
 
+const int FREELISTSIZE = 21;
```
Comment (fegin) [inline include/mxnet/mem_mgr.h]: FREE_LIST_SIZE. Without "_", it is not easy to read.

File: include/mxnet/mem_mgr.h [hunk 3]
```diff
@@ -1,17 +1,81 @@
...
+//min allocation size = 2^7 = 128 bytes
+inline int getFreeListIdx(size_t size) {
+  int idx = 0;
+  while (size > 0) {
+    size >>= 1;
+    idx++;
+  }  
+  idx -= 7;
+  idx = (idx < 0) ? 0 : idx;
+  idx = (idx > 19) ? 19 : idx;
```
Comment (fegin) [inline include/mxnet/mem_mgr.h]: 19 is kind of a magic number. You should explain.

File: include/mxnet/mem_mgr.h [hunk 4]
```diff
@@ -1,17 +1,81 @@
...
+    char* data_;
+    std::size_t size_;
+    Block* nextBlock_;
+    bool isFree_;
+    bool isHead_;
+
+  public:
+    Block(char* data, size_t size)
+      : data_(data)
+      , size_(size)
```
Comment (fegin) [inline include/mxnet/mem_mgr.h]: "," should be immediately after ")".

File: include/mxnet/mem_mgr.h [hunk 5]
```diff
@@ -1,17 +1,81 @@
...
+      , nextBlock_(NULL)
```
Comment (fegin) [inline include/mxnet/mem_mgr.h]: "," should be immediately after ")".

File: include/mxnet/mem_mgr.h [hunk 6]
```diff
@@ -1,17 +1,81 @@
...
+      , isFree_(true)
```
Comment (fegin) [inline include/mxnet/mem_mgr.h]: same here

File: include/mxnet/mem_mgr.h [hunk 7]
```diff
@@ -1,17 +1,81 @@
...
+      , isHead_(false) {
```
Comment (fegin) [inline include/mxnet/mem_mgr.h]: and here

**Review by fegin — CHANGES_REQUESTED** [review]
Feel free to merge after you fix my reviews.

**Pull request merged**